File: sdk/unreal/Source/ChillObservability/Private/ChillStore.cpp
// Copyright Chill Contributors. Licensed under the Apache License, Version 2.0.

#include "ChillStore.h"

#include "ChillIds.h"
#include "HAL/FileManager.h"
#include "HAL/PlatformFileManager.h"
#include "Misc/FileHelper.h"
#include "Misc/Paths.h"

namespace
{
	const TCHAR* const QueueMagic = TEXT("CHILLQ1\n");

	FString PaddedNanoseconds(const FString& Value)
	{
		FString Padded = Value;
		while (Padded.Len() < 19)
		{
			Padded = TEXT("0") + Padded;
		}
		return Padded;
	}
}

const TCHAR* FChillFileStore::GetMagic()
{
	return QueueMagic;
}

bool FChillMemoryStore::Append(const FChillQueuedRecord& Record)
{
	Records.Add(Record);
	return true;
}

TArray<FChillQueuedRecord> FChillMemoryStore::Peek(int32 MaximumRecords)
{
	TArray<FChillQueuedRecord> Selected;
	const int32 Count = FMath::Min(MaximumRecords, Records.Num());
	for (int32 Index = 0; Index < Count; ++Index)
	{
		Selected.Add(Records[Index]);
	}
	return Selected;
}

void FChillMemoryStore::Acknowledge(const TArray<FChillQueuedRecord>& Acknowledged)
{
	const int32 Count = FMath::Min(Acknowledged.Num(), Records.Num());
	if (Count > 0)
	{
		Records.RemoveAt(0, Count);
	}
}

void FChillMemoryStore::Purge()
{
	Records.Reset();
}

FChillFileStore::FChillFileStore(
	const FString& InDirectory,
	int64 InMaximumBytes,
	TFunction<void(const FString&)> InDiagnostic)
	: Directory(InDirectory)
	, MaximumBytes(InMaximumBytes)
	, Diagnostic(MoveTemp(InDiagnostic))
{
	IFileManager::Get().MakeDirectory(*Directory, true);
	Recover();
}

bool FChillFileStore::Append(const FChillQueuedRecord& Record)
{
	const FString Payload = FString(QueueMagic) + Record.LogJson;
	FTCHARToUTF8 Converted(*Payload);
	const int64 ByteCount = Converted.Length();
	if (CurrentBytes + ByteCount > MaximumBytes)
	{
		Diagnostic(TEXT("queue_full"));
		return false;
	}

	const FString Prefix = PaddedNanoseconds(Record.OccurredAtUnixNanoseconds)
		+ TEXT("-")
		+ FString::Printf(TEXT("%016lld"), Record.Sequence)
		+ TEXT("-")
		+ Record.RecordId;
	const FString FinalPath = FPaths::Combine(Directory, Prefix + TEXT(".json"));
	const FString TemporaryPath = FPaths::Combine(Directory, TEXT(".") + Prefix + TEXT(".tmp"));

	// Write to a temporary file first, then move it into place, so a crash mid-write
	// never leaves a partially written record in the readable queue.
	if (!FFileHelper::SaveStringToFile(
			Payload,
			*TemporaryPath,
			FFileHelper::EEncodingOptions::ForceUTF8WithoutBOM))
	{
		IFileManager::Get().Delete(*TemporaryPath, false, true, true);
		Diagnostic(TEXT("storage_failed"));
		return false;
	}
	if (!IFileManager::Get().Move(*FinalPath, *TemporaryPath, true, true))
	{
		IFileManager::Get().Delete(*TemporaryPath, false, true, true);
		Diagnostic(TEXT("storage_failed"));
		return false;
	}
	CurrentBytes += ByteCount;
	return true;
}

TArray<FChillQueuedRecord> FChillFileStore::Peek(int32 MaximumRecords)
{
	TArray<FChillQueuedRecord> Selected;
	TArray<FString> Files;
	IFileManager::Get().FindFiles(Files, *(Directory / TEXT("*.json")), true, false);
	Files.Sort();

	for (int32 Index = 0; Index < Files.Num() && Selected.Num() < MaximumRecords; ++Index)
	{
		const FString FullPath = FPaths::Combine(Directory, Files[Index]);
		FString Body;
		if (!FFileHelper::LoadFileToString(Body, *FullPath))
		{
			Quarantine(FullPath, true);
			continue;
		}
		if (!Body.StartsWith(QueueMagic, ESearchCase::CaseSensitive))
		{
			Quarantine(FullPath, true);
			continue;
		}
		const FString Name = FPaths::GetBaseFilename(Files[Index]);
		TArray<FString> Parts;
		Name.ParseIntoArray(Parts, TEXT("-"), false);
		// occurredAt-sequence-<uuid with four dashes> => at least six segments.
		if (Parts.Num() < 6)
		{
			Quarantine(FullPath, true);
			continue;
		}
		const int64 Sequence = FCString::Atoi64(*Parts[1]);
		FString RecordId = Parts[2];
		for (int32 PartIndex = 3; PartIndex < Parts.Num(); ++PartIndex)
		{
			RecordId += TEXT("-") + Parts[PartIndex];
		}
		Selected.Emplace(
			RecordId,
			Sequence,
			Parts[0],
			Body.RightChop(FCString::Strlen(QueueMagic)),
			FullPath,
			IFileManager::Get().FileSize(*FullPath));
	}
	return Selected;
}

void FChillFileStore::Acknowledge(const TArray<FChillQueuedRecord>& Records)
{
	for (const FChillQueuedRecord& Record : Records)
	{
		// Only delete inside the queue directory this store owns.
		if (Record.Path.IsEmpty() || FPaths::GetPath(Record.Path) != Directory)
		{
			continue;
		}
		if (IFileManager::Get().Delete(*Record.Path, false, true, true))
		{
			CurrentBytes = FMath::Max<int64>(0, CurrentBytes - Record.StoredBytes);
		}
		else
		{
			Diagnostic(TEXT("storage_failed"));
		}
	}
}

void FChillFileStore::Purge()
{
	TArray<FString> Files;
	IFileManager::Get().FindFiles(Files, *(Directory / TEXT("*")), true, false);
	for (const FString& File : Files)
	{
		IFileManager::Get().Delete(*FPaths::Combine(Directory, File), false, true, true);
	}
	CurrentBytes = 0;
}

void FChillFileStore::Recover()
{
	TArray<FString> Temporary;
	IFileManager::Get().FindFiles(Temporary, *(Directory / TEXT("*.tmp")), true, false);
	for (const FString& File : Temporary)
	{
		IFileManager::Get().Delete(*FPaths::Combine(Directory, File), false, true, true);
	}

	TArray<FString> Files;
	IFileManager::Get().FindFiles(Files, *(Directory / TEXT("*.json")), true, false);
	for (const FString& File : Files)
	{
		const FString FullPath = FPaths::Combine(Directory, File);
		FString Body;
		if (!FFileHelper::LoadFileToString(Body, *FullPath)
			|| !Body.StartsWith(QueueMagic, ESearchCase::CaseSensitive))
		{
			Quarantine(FullPath, false);
			continue;
		}
		CurrentBytes += IFileManager::Get().FileSize(*FullPath);
	}
}

void FChillFileStore::Quarantine(const FString& Path, bool bWasCounted)
{
	int64 StoredBytes = 0;
	if (bWasCounted)
	{
		StoredBytes = FMath::Max<int64>(0, IFileManager::Get().FileSize(*Path));
	}

	FString Target = Path + TEXT(".corrupt");
	if (IFileManager::Get().FileExists(*Target))
	{
		Target += TEXT(".") + FString::Printf(TEXT("%lld"), FDateTime::UtcNow().ToUnixTimestamp() * 1000);
	}
	if (IFileManager::Get().Move(*Target, *Path, true, true))
	{
		Diagnostic(TEXT("queue_corrupt"));
	}
	else
	{
		IFileManager::Get().Delete(*Path, false, true, true);
		Diagnostic(TEXT("storage_failed"));
	}
	CurrentBytes = FMath::Max<int64>(0, CurrentBytes - StoredBytes);
}

namespace ChillInstallation
{
	FString LoadOrCreate(const FString& QueueDirectory, const FString& Configured)
	{
		if (!Configured.IsEmpty())
		{
			return Configured;
		}

		const FString Parent = FPaths::GetPath(QueueDirectory.TrimEnd());
		const FString Directory = Parent.IsEmpty() ? QueueDirectory : Parent;
		IFileManager::Get().MakeDirectory(*Directory, true);
		const FString Path = FPaths::Combine(Directory, TEXT("installation-id"));

		FString Existing;
		if (FFileHelper::LoadFileToString(Existing, *Path))
		{
			Existing.TrimStartAndEndInline();
			if (ChillIds::IsUuidV4(Existing))
			{
				return Existing;
			}
		}

		const FString Generated = ChillIds::NewUuidV4();
		const FString Temporary = Path + TEXT(".tmp");
		if (FFileHelper::SaveStringToFile(
				Generated + TEXT("\n"),
				*Temporary,
				FFileHelper::EEncodingOptions::ForceUTF8WithoutBOM))
		{
			if (!IFileManager::Get().Move(*Path, *Temporary, true, true))
			{
				IFileManager::Get().Delete(*Temporary, false, true, true);
			}
		}
		return Generated;
	}
}

File: sdk/unity/Runtime/ChillStore.cs
using System;
using System.Collections.Generic;
using System.Globalization;
using System.IO;
using System.Text;

namespace Chill.Unity
{
    internal sealed class ChillQueuedRecord
    {
        internal ChillQueuedRecord(
            string recordId,
            long sequence,
            string occurredAtUnixNanoseconds,
            string logJson,
            string path = null,
            long storedBytes = 0)
        {
            RecordId = recordId;
            Sequence = sequence;
            OccurredAtUnixNanoseconds = occurredAtUnixNanoseconds;
            LogJson = logJson;
            Path = path;
            StoredBytes = storedBytes;
        }

        internal string RecordId { get; private set; }
        internal long Sequence { get; private set; }
        internal string OccurredAtUnixNanoseconds { get; private set; }
        internal string LogJson { get; private set; }
        internal string Path { get; private set; }
        internal long StoredBytes { get; private set; }
    }

    internal interface IChillStore
    {
        bool Append(ChillQueuedRecord record);
        IReadOnlyList<ChillQueuedRecord> Peek(int maximumRecords);
        void Acknowledge(IReadOnlyList<ChillQueuedRecord> records);
        void Purge();
    }

    internal sealed class ChillMemoryStore : IChillStore
    {
        private readonly List<ChillQueuedRecord> records = new List<ChillQueuedRecord>();

        public bool Append(ChillQueuedRecord record)
        {
            records.Add(record);
            return true;
        }

        public IReadOnlyList<ChillQueuedRecord> Peek(int maximumRecords)
        {
            int count = Math.Min(maximumRecords, records.Count);
            return records.GetRange(0, count);
        }

        public void Acknowledge(IReadOnlyList<ChillQueuedRecord> acknowledged)
        {
            int count = Math.Min(acknowledged.Count, records.Count);
            records.RemoveRange(0, count);
        }

        public void Purge()
        {
            records.Clear();
        }
    }

    internal sealed class ChillFileStore : IChillStore
    {
        private const string Magic = "CHILLQ1\n";
        private readonly string directory;
        private readonly long maximumBytes;
        private readonly Action<string> diagnostic;
        private long currentBytes;

        internal ChillFileStore(string directory, long maximumBytes, Action<string> diagnostic)
        {
            this.directory = directory;
            this.maximumBytes = maximumBytes;
            this.diagnostic = diagnostic;
            Directory.CreateDirectory(directory);
            Recover();
        }

        public bool Append(ChillQueuedRecord record)
        {
            byte[] bytes = Encoding.UTF8.GetBytes(Magic + record.LogJson);
            if (currentBytes + bytes.LongLength > maximumBytes)
            {
                diagnostic("queue_full");
                return false;
            }

            string prefix = record.OccurredAtUnixNanoseconds.PadLeft(19, '0')
                + "-"
                + record.Sequence.ToString("D16", CultureInfo.InvariantCulture)
                + "-"
                + record.RecordId;
            string finalPath = System.IO.Path.Combine(directory, prefix + ".json");
            string temporaryPath = System.IO.Path.Combine(directory, "." + prefix + ".tmp");
            try
            {
                using (var stream = new FileStream(
                    temporaryPath,
                    FileMode.CreateNew,
                    FileAccess.Write,
                    FileShare.None,
                    4096,
                    FileOptions.None))
                {
                    stream.Write(bytes, 0, bytes.Length);
                    stream.Flush();
                    try
                    {
                        stream.Flush(true);
                    }
                    catch (PlatformNotSupportedException)
                    {
                        // WebGL and some console filesystems expose only the virtual flush.
                    }
                }
                File.Move(temporaryPath, finalPath);
                currentBytes += bytes.LongLength;
                return true;
            }
            catch
            {
                TryDelete(temporaryPath);
                diagnostic("storage_failed");
                return false;
            }
        }

        public IReadOnlyList<ChillQueuedRecord> Peek(int maximumRecords)
        {
            var selected = new List<ChillQueuedRecord>();
            string[] files;
            try
            {
                files = Directory.GetFiles(directory, "*.json", SearchOption.TopDirectoryOnly);
                Array.Sort(files, StringComparer.Ordinal);
            }
            catch
            {
                diagnostic("storage_failed");
                return selected;
            }

            for (int index = 0; index < files.Length && selected.Count < maximumRecords; index += 1)
            {
                try
                {
                    string body = File.ReadAllText(files[index], Encoding.UTF8);
                    if (!body.StartsWith(Magic, StringComparison.Ordinal))
                    {
                        Quarantine(files[index], true);
                        continue;
                    }
                    string name = System.IO.Path.GetFileNameWithoutExtension(files[index]);
                    string[] parts = name.Split(new[] { '-' }, 4);
                    long sequence = 0;
                    if (parts.Length != 4 || !long.TryParse(parts[1], NumberStyles.None, CultureInfo.InvariantCulture, out sequence))
                    {
                        Quarantine(files[index], true);
                        continue;
                    }
                    var info = new FileInfo(files[index]);
                    selected.Add(new ChillQueuedRecord(
                        parts[2] + "-" + parts[3],
                        sequence,
                        parts[0],
                        body.Substring(Magic.Length),
                        files[index],
                        info.Length));
                }
                catch
                {
                    Quarantine(files[index], true);
                }
            }
            return selected;
        }

        public void Acknowledge(IReadOnlyList<ChillQueuedRecord> records)
        {
            for (int index = 0; index < records.Count; index += 1)
            {
                ChillQueuedRecord record = records[index];
                if (string.IsNullOrEmpty(record.Path)
                    || !string.Equals(System.IO.Path.GetDirectoryName(record.Path), directory, StringComparison.Ordinal))
                {
                    continue;
                }
                try
                {
                    File.Delete(record.Path);
                    currentBytes = Math.Max(0, currentBytes - record.StoredBytes);
                }
                catch
                {
                    diagnostic("storage_failed");
                }
            }
        }

        public void Purge()
        {
            try
            {
                string[] files = Directory.GetFiles(directory, "*", SearchOption.TopDirectoryOnly);
                for (int index = 0; index < files.Length; index += 1)
                {
                    TryDelete(files[index]);
                }
                currentBytes = 0;
            }
            catch
            {
                diagnostic("storage_failed");
            }
        }

        private void Recover()
        {
            try
            {
                string[] temporary = Directory.GetFiles(directory, "*.tmp", SearchOption.TopDirectoryOnly);
                for (int index = 0; index < temporary.Length; index += 1)
                {
                    TryDelete(temporary[index]);
                }

                string[] files = Directory.GetFiles(directory, "*.json", SearchOption.TopDirectoryOnly);
                for (int index = 0; index < files.Length; index += 1)
                {
                    try
                    {
                        string body = File.ReadAllText(files[index], Encoding.UTF8);
                        if (!body.StartsWith(Magic, StringComparison.Ordinal))
                        {
                            Quarantine(files[index], false);
                            continue;
                        }
                        currentBytes += new FileInfo(files[index]).Length;
                    }
                    catch
                    {
                        Quarantine(files[index], false);
                    }
                }
            }
            catch
            {
                diagnostic("storage_failed");
            }
        }

        private void Quarantine(string path, bool wasCounted)
        {
            long storedBytes = 0;
            if (wasCounted)
            {
                try
                {
                    storedBytes = new FileInfo(path).Length;
                }
                catch
                {
                    // The record still leaves the readable queue below.
                }
            }
            try
            {
                string target = path + ".corrupt";
                if (File.Exists(target))
                {
                    target += "." + DateTimeOffset.UtcNow.ToUnixTimeMilliseconds().ToString(CultureInfo.InvariantCulture);
                }
                File.Move(path, target);
                diagnostic("queue_corrupt");
            }
            catch
            {
                TryDelete(path);
                diagnostic("storage_failed");
            }
            currentBytes = Math.Max(0, currentBytes - storedBytes);
        }

        private static void TryDelete(string path)
        {
            try
            {
                if (File.Exists(path))
                {
                    File.Delete(path);
                }
            }
            catch
            {
                // The bounded diagnostic at the caller is sufficient.
            }
        }
    }

    internal static class ChillInstallation
    {
        internal static string LoadOrCreate(string queueDirectory, string configured)
        {
            if (!string.IsNullOrEmpty(configured))
            {
                return configured;
            }

            string parent;
            string path;
            try
            {
                parent = Directory.GetParent(queueDirectory) == null
                    ? queueDirectory
                    : Directory.GetParent(queueDirectory).FullName;
                Directory.CreateDirectory(parent);
                path = System.IO.Path.Combine(parent, "installation-id");
                if (File.Exists(path))
                {
                    string existing = File.ReadAllText(path, Encoding.UTF8).Trim();
                    if (ChillIds.IsUuidV4(existing))
                    {
                        return existing;
                    }
                }
            }
            catch
            {
                return ChillIds.NewUuidV4();
            }

            string generated = ChillIds.NewUuidV4();
            string temporary = path + ".tmp";
            try
            {
                File.WriteAllText(temporary, generated + "\n", Encoding.UTF8);
                if (File.Exists(path))
                {
                    File.Delete(path);
                }
                File.Move(temporary, path);
            }
            catch
            {
                try
                {
                    if (File.Exists(temporary))
                    {
                        File.Delete(temporary);
                    }
                }
                catch
                {
                    // The generated identity remains valid for this process.
                }
            }
            return generated;
        }
    }
}
